java/vector/src/test/java/org/apache/arrow/vector/TestBitVectorHelper.java

```diff
@@ -222,6 +223,59 @@ public void testConcatBitsInPlace() {
     }
   }
 
+  @Test
+  public void testLoadValidityBuffer() {
+    try (RootAllocator allocator = new RootAllocator(1024)) {
+      // if the input validity buffer is all null, we should allocate new memory
+      ArrowFieldNode fieldNode = new ArrowFieldNode(1024, 1024);
+      try (ArrowBuf buf = BitVectorHelper.loadValidityBuffer(fieldNode, null, allocator)) {
+        assertEquals(128, allocator.getAllocatedMemory());
+        for (int i = 0; i < 128; i++) {
+          assertEquals(0, buf.getByte(i));
+        }
+      }
+
+      // should also allocate memory if input validity buffer is all not-null
+      fieldNode = new ArrowFieldNode(1024, 0);
+      try (ArrowBuf buf = BitVectorHelper.loadValidityBuffer(fieldNode, null, allocator)) {
+        assertEquals(128, allocator.getAllocatedMemory());
+        for (int i = 0; i < 128; i++) {
+          assertEquals((byte) 0xff, buf.getByte(i));
+        }
+      }
+
+      // should not allocate memory if input validity buffer is not null, even if all values are
+      // null
+      fieldNode = new ArrowFieldNode(1024, 1024);
+      try (ArrowBuf src = allocator.buffer(128);
+           ArrowBuf dst = BitVectorHelper.loadValidityBuffer(fieldNode, src, allocator)) {
+        assertEquals(128, allocator.getAllocatedMemory());
+      }
+
+      // ... similarly if all values are not null
+      fieldNode = new ArrowFieldNode(1024, 0);
+      try (ArrowBuf src = allocator.buffer(128);
+           ArrowBuf dst = BitVectorHelper.loadValidityBuffer(fieldNode, src, allocator)) {
+        assertEquals(128, allocator.getAllocatedMemory());
+      }
+
+      // mixed case, input should match output
+      int numNulls = 100;
+      fieldNode = new ArrowFieldNode(1024, numNulls);
+      try (ArrowBuf src = allocator.buffer(128)) {
+        for (int i = 0; i < numNulls; i++) {
+          BitVectorHelper.setBit(src, i);
+        }
+        try (ArrowBuf dst = BitVectorHelper.loadValidityBuffer(fieldNode, src, allocator)) {
+          assertEquals(128, allocator.getAllocatedMemory());
```

Comment:
We can assert that `src` and `dst` refer to the same underlying buffer?

Reply:
We can't compare by reference since `dst` is a different `ArrowBuf` object (see `ReferenceManager.retain`). I added a comparison on memory address of these two.